### PR TITLE
Fix ext_proc integration test fuzzer issue

### PR DESCRIPTION
**Pull request by yanjunxiang-google**
This is to fix the ext_proc fuzzer test issue: https://oss-fuzz.com/testcase-detail/5087529639739392

The root cause is the test server is accessing some local variables out of its lifetime.  

The solution is to changing them to static.

BTW, the issue is a fuzzer specific issue.

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz.h [hunk 1]
```diff
@@ -229,8 +229,9 @@ fuzzExtProcRun(const test::extensions::filters::http::ext_proc::ExtProcGrpcTestC
   FuzzedDataProvider downstream_provider(
       reinterpret_cast<const uint8_t*>(input.downstream_data().data()),
       input.downstream_data().size());
-  FuzzedDataProvider ext_proc_provider(
-      reinterpret_cast<const uint8_t*>(input.ext_proc_data().data()), input.ext_proc_data().size());
+  static std::string ext_proc_data = input.ext_proc_data();
+  static FuzzedDataProvider ext_proc_provider(
```
[the remark this answers is not in the document]
Comment (adisuissa) [reply test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz.h:233]: Correct me if I'm wrong, but isn't `ext_proc_data` and `ext_proc_provider` dependent on the fuzz test input?
If so, why should this be static?

[the remark this answers is not in the document]
Comment (yanjunxiang-google) [reply test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz.h:233]: Thanks for review.

The ext_proc_provider address is passed into the c'tor of a persistent static object: fuzz_helper:

https://github.com/envoyproxy/envoy/blob/73da4c3a3fdd11192ec9f8dc60906e2f3402132d/test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz.h#L245

https://github.com/envoyproxy/envoy/blob/0bc754a213b6cd28c7117c82de0f721bb19d6ef2/test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz_helper.cc#L106

And the ext_proc_provider is used by the persistent test server here: https://github.com/envoyproxy/envoy/blob/0bc754a213b6cd28c7117c82de0f721bb19d6ef2/test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz_helper.cc#L451


Thus ext_proc_provider has to be persistent. And ext_proc_data is a pointer to construct ext_proc_provider so it has to be persistent as well.

In a nutshell, ext_proc_provider is an address used by the persistent fuzz_helper, thus it has to be persistent across different test runs.   Thus the solution is to copy the test input into this static variable and using it to construct fuzz_helper.

[the remark this answers is not in the document]
Comment (adisuissa) [reply test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz.h:233]: So I guess either the fuzz_helper object should not be static, or the data should be explicitly updated on each run.

[the remark this answers is not in the document]
Comment (yanjunxiang-google) [reply test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz.h:233]: Test server is using fuzz_helper to send random response. And we want to have test server persistent to avoid launch it each run.

So, by  copy the test input into this static variable and using it to construct fuzz_helper, the ext_proc server data is updated during each run.

And the downstream_provider, is another part of data derived from the test input. This data is not used to launch Envoy or Test server (both of them are persistent), thus it is non-static:

https://github.com/envoyproxy/envoy/blob/0bc754a213b6cd28c7117c82de0f721bb19d6ef2/test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz.h#L229

https://github.com/envoyproxy/envoy/blob/0bc754a213b6cd28c7117c82de0f721bb19d6ef2/test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz.h#L277

[the remark this answers is not in the document]
Comment (adisuissa) [reply test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz.h]: As discussed offline, I'm still not convinced that the fuzzer will ensure that different iterations will not try to access old/new state objects.
I wonder which static objects make the persistent fuzzer much faster than the non-persistent variant. Can you provide a breakdown of the different objects initialization costs?

[the remark this answers is not in the document]
Comment (yanjunxiang-google) [reply test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz.h]: Sorry for not being able to respond early.
With below code:
  ext_proc_data = std::make_unique<std::string>(input.ext_proc_data());

ext_proc_data is refreshed after each iteration.  I dumped the contents of ext_proc_data and confirmed it matching with input in each iteration.

Regarding the cost of initialize Envoy and test-server,  I did some measurements when we add this persistent fuzzer. I recall it's about ~450ms to start Envoy and ~80ms to start  the test server in my cloud top VM during each iteration.  The persistent fuzzer is just trying to amortize that in 2000 iterations.

[the remark this answers is not in the document]
Comment (adisuissa) [reply test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz.h]: > ext_proc_data is refreshed after each iteration. I dumped the contents of ext_proc_data and confirmed it matching with input in each iteration.

IIUC this is a flaky test, and so there may be a case where they will not be matched.
I'm going to approve this, but I suggest considering refactoring to use a FakeUpstream, and solve the underlying issue if this persists.

**Comment by adisuissa** [comment]
Can you please merge main again?
